lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,397 @@
+"""
+Process presence-absence tests into the relational warehouse
+
+The ETL process performs a `find_or_create` for targets, as we anticipate the same targets to be tested, but leave room for the addition of new targets.
+
+The ETL process performs a `find_or_create` for samples to allow the processing of re-tested samples. We do not expect to update information of a sample itself.
+
+The ETL process performs an `upsert` for presence_absence to allow the update of results should there be a re-test on an old sample.
+
+The presence absence ETL process will abort under these conditions:
+1. If a sample's barcode matches with a UUID that is of the incorrect identifier set
+2. If we receive an unexpected value for the "controlStatus" of a target
+3. If we receive an unexpected value for the "targetResult" of a specific test
+"""
+import click
+import logging
+from datetime import datetime, timezone
+from typing import Any
+from seattleflu.db.session import DatabaseSession
+from seattleflu.db.datatypes import Json
+from . import etl
+
+
+LOG = logging.getLogger(__name__)
+
+
+# This revision number is stored in the processing_log of each presence-absence
+# record when the presence-absence test is successfully processed by this ETL 
+# routine. The routine finds new-to-it records to process by looking for
+# presence-absence tests lacking this revision number in their log.  If a 
+# change to the ETL routine necessitates re-processing all presence-absence tests, 
+# this revision number should be incremented.
+REVISION = 1
+
+
+@etl.command("presence-absence", help = __doc__)
+
+@click.option("--dry-run", "action",
+    help        = "Only go through the motions of changing the database (default)",
+    flag_value  = "rollback",
+    default     = True)
+
+@click.option("--prompt", "action",
+    help        = "Ask if changes to the database should be saved",
+    flag_value  = "prompt")
+
+@click.option("--commit", "action",
+    help        = "Save changes to the database",
+    flag_value  = "commit")
+
+def etl_presence_absence(*, action: str):
+    LOG.debug(f"Starting the presence_absence ETL routine, revision {REVISION}")
+
+    db = DatabaseSession()
+
+    # Fetch and iterate over presence-absence tests that aren't processed
+    #
+    # Use a server-side cursor by providing a name.  This ensures we limit how
+    # much data we fetch at once, to limit local process size.  Each presence-
+    # absence test document is ~10 KB and the default fetch size 
+    # (cursor.itersize) is 2,000, thus we'll get ~20 MB on each fetch of 2,000 
+    # presence-absence test.
```

Comment:
While this now uses the right noun (presence-absence), it still isn't accurate. The document sizes and total fetch size aren't the same as enrollments. Either the comment should be made accurate, or it should go. :-)